File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.py
import sys
from collections import deque

input = sys.stdin.readline

n, m, k = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
operator = [list(map(int, input().split())) for _ in range(k)]
operator = [[r-1, c-1, s] for r, c, s in operator]
    
def rotate(A, r, c, s):
    for layer in range(1, s+1):
        top, bottom = r-layer, r+layer
        left, right = c-layer, c+layer
        
        nxt = A[top][left]
        for i in range(left+1, right+1):
            A[top][i], nxt = nxt, A[top][i]
        
        for i in range(top+1, bottom+1):
            A[i][right], nxt = nxt, A[i][right]
        
        for i in range(right-1, left-1, -1):
            A[bottom][i], nxt = nxt, A[bottom][i]
            
        for i in range(bottom-1, top-1, -1):
            A[i][left], nxt = nxt, A[i][left]

ans = 50 * 100 + 1
stack = deque()
def bt():
    global ans, k
    
    if len(stack) == k:
        A = [line[:] for line in arr]
        for x in stack:
            r, c, s = operator[x]
            rotate(A, r, c, s)
        
        ans = min(ans, min(map(sum, A)))
        return
    
    for i in range(k):
        if not i in stack:
            stack.append(i)
            bt()
            stack.pop()
    
    if len(stack) < k:
        return

bt()    
print(ans)

File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int n, m, k;
    static int[][] inputArr;
    static int[][] command;

    static int ans;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        inputArr = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                inputArr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        command = new int[k][3];
        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 2; j++) {
                command[i][j] = Integer.parseInt(st.nextToken())-1;
            }
            command[i][2] = Integer.parseInt(st.nextToken());
        }

        ans = Integer.MAX_VALUE;
        visited = new boolean[k];
        bt(deepcopy(inputArr), 0);
        System.out.println(ans);
    }

    static void bt(int[][] arr, int depth) {
        if (depth == k) {
            for (int[] line: arr) {
                int total = 0;
                for (int num: line) {
                    total += num;
                }
                ans = Math.min(ans, total);
            }

            return;
        }

        for (int i = 0; i < k; i++) {
            if (!visited[i]) {
                visited[i] = true;
                int[][] newArr = deepcopy(arr);
                rotate(newArr, i);
                bt(newArr, depth+1);
                visited[i] = false;
            }
        }
    }

    static int[][] deepcopy(int[][] A) {
        int[][] copy = new int[n][m];
        for (int i = 0; i < n; i++) {
            System.arraycopy(A[i], 0, copy[i], 0, m);
        }

        return copy;
    }

    static void rotate(int[][] arr, int idx) {
        int r = command[idx][0], c = command[idx][1], s = command[idx][2];

        for (int layer=1; layer<=s; layer++) {
            int top = r-layer, bottom = r+layer;
            int left = c-layer, right = c+layer;

            int save = arr[top][left];
            for (int j=left+1; j<=right; j++) {
                int tmp = arr[top][j];
                arr[top][j] = save;
                save = tmp;
            }

            for (int i=top+1; i<=bottom; i++) {
                int tmp = arr[i][right];
                arr[i][right] = save;
                save = tmp;
            }

            for (int j=right-1; j>=left; j--) {
                int tmp = arr[bottom][j];
                arr[bottom][j] = save;
                save = tmp;
            }

            for (int i=bottom-1; i>=top; i--) {
                int tmp = arr[i][left];
                arr[i][left] = save;
                save = tmp;
            }

        }

    }

}
